packages/react-components/react-datepicker-compat/src/components/DatePicker/useDatePicker.tsx

```diff
@@ -358,24 +362,31 @@ export const useDatePicker_unstable = (props: DatePickerProps, ref: React.Ref<HT
       contentAfter: <CalendarMonthRegular onClick={onIconClick as unknown as React.MouseEventHandler<SVGElement>} />,
       readOnly: !allowTextInput,
       role: 'combobox',
-      root: {
-        'aria-owns': open ? popupSurfaceId : undefined,
-        ref: useMergedRefs(triggerWrapperRef, ref),
-      },
-      input: {
-        ref: rootRef,
-      },
     },
   });
 
-  rootShorthand.onChange = mergeCallbacks(rootShorthand.onChange, onInputChange);
-  rootShorthand.onBlur = mergeCallbacks(rootShorthand.onBlur, onInputBlur);
-  rootShorthand.onKeyDown = mergeCallbacks(rootShorthand.onKeyDown, onInputKeyDown);
-  rootShorthand.onFocus = mergeCallbacks(rootShorthand.onFocus, onInputFocus);
-  rootShorthand.onClick = mergeCallbacks(rootShorthand.onClick, onInputClick);
+  const rootShorthand = resolveShorthand(props.root, {
+    required: true,
+    defaultProps: {
+      'aria-owns': open ? popupSurfaceId : undefined,
+      ref: triggerWrapperRef,
+    },
+  });
+  rootShorthand.ref = useMergedRefs(rootShorthand.ref, triggerWrapperRef);
+  root.root = rootShorthand;
+
+  const inputShorthand = resolveShorthand(props.input, { required: true });
+  inputShorthand.ref = useMergedRefs(inputShorthand.ref, ref, rootRef);
+  root.input = inputShorthand;
+
+  root.onChange = mergeCallbacks(root.onChange, onInputChange);
+  root.onBlur = mergeCallbacks(root.onBlur, onInputBlur);
+  root.onKeyDown = mergeCallbacks(root.onKeyDown, onInputKeyDown);
+  root.onFocus = mergeCallbacks(root.onFocus, onInputFocus);
+  root.onClick = mergeCallbacks(root.onClick, onInputClick);
```

Comment:
mergeCallbacks doesn't memoize its result, so I think these will be new instances on every render, which undoes all of the careful use of `React.useCallback` above.

This may be better done in a separate PR, but I would recommend removing all uses of `React.useCallback` in this function, and instead wrapping these with our `useEventCallback` function, which provides a stable function reference even when its dependencies change on every render:
```ts
root.onChange = useEventCallback(mergeCallbacks(root.onChange, onInputChange));
// etc.
```

Reply:
it's a small enough change, so I'll add it as well